Add RTDS and ADS apis to mobile engine builder

caschoener opened the pull request:
Updating engine_builder to support rtds and ads configuration.  Currently ads always uses google_grpc, so updated the test flags to compile that in.

caschoener commented on mobile/test/common/integration/base_client_integration_test.cc:
```diff
@@ -194,9 +194,8 @@ void BaseClientIntegrationTest::createEnvoy() {
 
-  finalizeConfigWithPorts(config_helper_, ports, use_lds_);
-
   if (override_builder_config_) {
```
`override_builder_config_` is always `false` now, so this is dead code.  But I remember Alyssa mentioned it might be good to leave it in in case we needed to use `override_builder_config_` again?

caschoener commented on mobile/test/common/integration/xds_integration_test.cc:
```diff
@@ -27,20 +72,17 @@ XdsIntegrationTest::XdsIntegrationTest() : BaseClientIntegrationTest(ipVersion()
+  // We use a fake port at first, and assign a real one during initialization
+  builder_.setCustomClusters(
+      xdsClustersConfig(Network::Test::getLoopbackAddressString(ipVersion()), 0, 0));
+  builder_.disableStatsConfig(true);
```
Reposting an unresolved comment from Alyssa in the other repo:

"do you know what stat counters we need for xDS? maybe we should we just add them into the default matchers now we're adding builder xDS support?"

Might need tips on how to do this

caschoener commented on mobile/test/common/integration/rtds_integration_test.cc:
```diff
@@ -72,8 +70,8 @@ TEST_P(RtdsIntegrationTest, RtdsReload) {
-  EXPECT_EQ(cc_.on_header_consumed_bytes_from_response, 13);
-  EXPECT_EQ(cc_.on_complete_received_byte_count, 41);
+  EXPECT_EQ(cc_.on_header_consumed_bytes_from_response, 27);
+  EXPECT_EQ(cc_.on_complete_received_byte_count, 67);
```
these changed because we switched to http 1 from https 2

caschoener commented on mobile/test/common/integration/sds_integration_test.cc:
```diff
@@ -19,9 +19,25 @@ const envoy::service::secret::v3::SdsDummy _sds_dummy;
 class SdsIntegrationTest : public XdsIntegrationTest {
 public:
   SdsIntegrationTest() {
+    override_builder_config_ = true;
```
moving some of the shared code from xds to here because the rtds test has switched to using the builder API

caschoener commented:
Think I will wait to add APIs for the other languages until someone says this looks good btw

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -498,6 +498,8 @@ stats_sinks: *stats_sinks
         - prefix: cluster.base.
         - prefix: cluster.base_h2.
+        - prefix: sds.
+        - prefix: runtime.
```
Nice!

Augustyniak replied:
How many stats does this effectively enable - as in, how many more are tracked by Envoy after that change (we can just look at the number of active stats after we open the example app I guess)? (also, you may need to rebase as `patterns:` were updated in main)

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +540,44 @@ stats_sinks: *stats_sinks
+            api_type: {}
+            transport_api_version: V3
+            grpc_services:
+              envoy_grpc:
```
While Envoy Mobile has google_grpc disabled by default, we will need to enable it for the Traffic Director testing amongst other use cases (some users may want to enable google_grpc for the same reasons).  So we probably also want to make this configurable through the APIs, or provide a separate API, e.g. addGoogleGrpcXdsCluster vs. addEnvoyGrpcXdsCluster.

alyssawilk replied:
I'm inclined to get Lyft's opinion on this but as currently we only have one use case for xDS which uses google's gRPC, I'd be inclined to as we add API docs, add an API doc saying that addWhateverWeCallThis requires building with --build_flags_for_google_grpc.  when/if folks want to do xDS with non-google-gRPC we could add a boolean to the function indicating which client should be used.

Augustyniak replied:
The solution proposed by Alyssa sounds good to me. It's fine to keep xds coupled with google grpc client as long as it's well documented behavior.

abeyad replied:
That sounds fine to me as well, but then, we'd have to update the integration tests to use google_grpc instead

caschoener replied:
Working on switching to google_grpc now

caschoener commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -189,6 +189,18 @@ EngineBuilder& EngineBuilder::enableH2ExtendKeepaliveTimeout(bool h2_extend_keep
+}
+
+EngineBuilder& EngineBuilder::addRtdsLayer(const std::string& api_type,
+                                           const std::string xds_cluster) {
```
Isn't everything const already? Guessing maybe you meant the int port?

RyanTheOptimist replied:
I think "const std::string xds_cluster" => "const std::string& xds_cluster".

caschoener commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +540,44 @@ stats_sinks: *stats_sinks
 // clang-format on
+
+const char* xds_cluster_insert = R"(
+    - name: some_static_layer
```
Sorry. I mixed up the variable names for the rtds and xds insert lol.  I think this comment still applies though, seems like we don't want this random some_static_layer in production? Maybe the best thing to do would be to change the test to only use dynamic layers:
1. send an RTDS request to set test_feature to True, then verify it's true
2. send RTDS request to set test_feature to False, then verify

abeyad replied:
is the variable only used in RTDS tests?  if so, i think it makes sense to move the flag to the dynamic layer.  in any case, we should consider adding a static runtime layer API to the engine builder.

caschoener commented on mobile/library/cc/engine_builder.h:
```diff
@@ -61,8 +61,13 @@ class EngineBuilder {
-
+  // Creates an XDS cluster with a single static endpoint, where the endpoint is intended to be of
+  // a fake upstream on the loopback address.
+  EngineBuilder& addXdsCluster(const std::string& address, int port);
```
Does adding a configurable name solve this? It does seem to me that some use cases might want to share xds clusters, or create new one per xds service.  Not sure if that's the case though.

abeyad commented on mobile/library/common/config/templates.h:
```diff
@@ -102,3 +102,13 @@ extern const char* default_cert_validation_context_template;
+/**
+ * Config template for an XDS cluster
+ */
+extern const char* xds_cluster_insert;
```
i think we probably want to change this name to `cluster_insert` and the builder API to `addCluster`

abeyad reviewed (commented):
I think this is looking more along the lines of what I was thinking!  Let's see what Alyssa thinks.  Thanks for the iterations on it.

alyssawilk reviewed (commented):
Sweet!  Some thoughts from my side since Ali's out

alyssawilk commented on mobile/library/cc/engine_builder.h:
```diff
@@ -61,7 +61,10 @@ class EngineBuilder {
   EngineBuilder& enablePlatformCertificatesValidation(bool platform_certificates_validation_on);
-
+  // Creates an XDS cluster with a single static endpoint
+  EngineBuilder& addCluster(const std::string& name, const std::string& address, int port);
```
oh, or update the comment if the insert is generic and not xds specific :-)

abeyad replied:
+1, the API is adding a generic cluster, it just so happens we'll use it in the `addRtdsLayer` API as well as other ones added in the future.  So I think `addCluster` makes the most sense.  Comment needs to be updated as Alyssa mentioned.

abeyad commented:
/wait

caschoener commented:
cool here's a new version.  Enabled  google_grpc for non-production use-cases in the bazelrc.  Switched to google_grpc in the test, so there's no longer an xds cluster. Switched from a test-specific static layer to just read/writing one of the existing params. 

Changes can be seen here https://github.com/envoyproxy/envoy/pull/24860/files/436d1bebe152c80326a6c0ee4d3c46d236e338ea..8a2a68dfa9bba5b86d809ffc86f0dd0af5a495f7

caschoener commented:
Latest commit adds the kotlin API.  I realized it's simplest to store all the params in the builder and then do the formatting later so let's go with that.

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -303,6 +313,13 @@ std::string EngineBuilder::generateConfigStr() const {
   }
 
+  if (!rtds_layer_name_.empty()) {
+    std::string custom_layers_ = fmt::format(rtds_layer_insert, rtds_layer_name_, rtds_layer_name_,
```
trailing underscores on variable names are only meant for class members, but see note above about making custom_layers_ the class variable instead of a local one

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +540,41 @@ stats_sinks: *stats_sinks
+                target_uri: "{}:{}"
+            set_node_on_first_message_only: true)";
+            
+const char* cluster_insert = R"(
```
don't think we're using this anymore; remove

abeyad replied:
you didn't remove this?

abeyad commented on mobile/library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:
```diff
@@ -571,6 +571,27 @@ open class EngineBuilder(
+  /**
+   * Add a google_grpc RTDS layer to config
+   *
+   * @param layer_name the layer name.
```
don't think we need the extra line of whitespace between each "@param"

abeyad commented on mobile/library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:
```diff
@@ -571,6 +571,27 @@ open class EngineBuilder(
+    this.rtds_address_ = address
+    this.rtds_port_ = port
+    return this
+  }
```
what is this doing exactly? the Kotlin functions would need to use a bridge to call the C++ APIs, no?  for starters, i think it needs to be passed into https://sourcegraph.com/github.com/envoyproxy/envoy@ec9099786796da6f834a6d562d0c3939c342a5e1/-/blob/mobile/library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt?L583, which would then go through some other layers till it's initializing the Engine creation.

Also, I don't think you declared the rtds_* variables in the Kotlin file, so this probably won't compile.

caschoener commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -189,6 +189,16 @@ EngineBuilder& EngineBuilder::enableH2ExtendKeepaliveTimeout(bool h2_extend_keep
+EngineBuilder& EngineBuilder::addRtdsLayer(const std::string& layer_name,
+                                           const std::string& api_type, const std::string& address,
+                                           const int port) {
+  rtds_layer_name_ = layer_name;
```
It seems like this would complicate the API for the other languages because I'd have to do a fmt::format equivalent in each language

caschoener replied:
I suppose the more I look at it it's going to be very tedious either way so IDK

caschoener commented:
After talking it over I think adding all the languages in one diff will be a lot, so let's just ship this one as-is.

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +540,41 @@ stats_sinks: *stats_sinks
+          initial_fetch_timeout:
+            seconds: 1
+          resource_api_version: V3
+          api_config_source:
```
it just occurred to me that for Traffic Director, it only supports ADS.  So we will also want to support ads instead of api_config_source: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/core/v3/config_source.proto#L225.  It would require setting the bootstrap's ads_config field: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/bootstrap/v3/bootstrap.proto#L101

I think we can start with only supporting ADS, and add non-ADS support later.  This will just require moving a bunch of this config stuff to `bootstrap.dynamic_resources.ads_config` and have `rtds_config` just set the `ads` field.

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -189,6 +189,17 @@ EngineBuilder& EngineBuilder::enableH2ExtendKeepaliveTimeout(bool h2_extend_keep
+  custom_layers_ = fmt::format(rtds_layer_insert, layer_name, layer_name);
+  return *this;
+}
+EngineBuilder& EngineBuilder::setAggregatedDiscoveryService(const std::string& api_type,
```
`api_type` will always be `GRPC`, at least for now, so we can further simply it

caschoener replied:
Ali changed his mind

abeyad commented on mobile/library/common/config/templates.h:
```diff
@@ -102,3 +102,18 @@ extern const char* default_cert_validation_context_template;
+/**
+ * Config template for a new cluster
+ */
+extern const char* cluster_insert;
```
need to remove this too

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -303,6 +314,16 @@ std::string EngineBuilder::generateConfigStr() const {
+  }
+
+  if (!custom_ads_.empty()) {
+    absl::StrReplaceAll({{"#{custom_ads}", absl::StrCat("#{custom_ads}\n", custom_ads_)}},
```
before setting custom_layers_, we should set custom_ads_, and then before setting custom_layers_, check that custom_ads_ is set, and if not, throw a std::runtime_error

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -184,6 +184,22 @@ EngineBuilder& EngineBuilder::enforceTrustChainVerification(bool trust_chain_ver
   return *this;
 }
 
+EngineBuilder& EngineBuilder::enableH2ExtendKeepaliveTimeout(bool h2_extend_keepalive_timeout_on) {
```
this shouldn't be added as part of this PR?

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -491,6 +492,25 @@ R"(
+  stats_matcher:
+    inclusion_list:
+      patterns:
+        - prefix: cluster.base.
```
seems like a lot more patterns were added, probably from a merge conflict?

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -302,6 +326,7 @@ std::string EngineBuilder::generateConfigStr() const {
   if (config_str.find("{{") != std::string::npos) {
     throw std::runtime_error("could not resolve all template keys in config");
   }
+  std::cout << config_str << std::endl;
```
please remove

abeyad commented:
Added @jpsim for cross-company review

abeyad commented on mobile/test/common/integration/rtds_integration_test.cc:
```diff
@@ -85,17 +68,16 @@ TEST_P(RtdsIntegrationTest, RtdsReload) {
+  EXPECT_TRUE(Runtime::runtimeFeatureEnabled(
+      "envoy.reloadable_features.skip_dns_lookup_for_proxied_requests"));
 }
 
```
Can we also add a test to ensure we throw an error when addRtdsLayer is called without calling setAggregatedDiscoveryService?

abeyad commented on mobile/library/cc/engine_builder.h:
```diff
@@ -60,6 +60,10 @@ class EngineBuilder {
   EngineBuilder& enablePlatformCertificatesValidation(bool platform_certificates_validation_on);
+  // Adds an RTDS layer to default config
+  EngineBuilder& addRtdsLayer(const std::string& layer_name);
+  EngineBuilder& setAggregatedDiscoveryService(const std::string& api_type,
```
Can we add a comment to this method?

abeyad commented on mobile/library/cc/engine_builder.h:
```diff
@@ -60,6 +60,10 @@ class EngineBuilder {
   EngineBuilder& enableDrainPostDnsRefresh(bool drain_post_dns_refresh_on);
   EngineBuilder& enforceTrustChainVerification(bool trust_chain_verification_on);
   EngineBuilder& enablePlatformCertificatesValidation(bool platform_certificates_validation_on);
+  // Adds an RTDS layer to default config
```
Can we also add that setAggregatedDiscoveryService must be called before addRtdsLayer is called?

caschoener commented:
/retest

jpsim commented:
Just an FYI according to the [mobile_perf CI job](https://github.com/envoyproxy/envoy/actions/runs/4008291380) this increases the binary size by ~1.4%. Not a blocker for me though, but raising it because reducing binary size have been an active effort in the last few months.

caschoener commented:
/retest

jpsim reviewed (dismissed):
Do you have a ticket to port this over to the Swift/Kotlin engine builders?

RyanTheOptimist commented:
> Just an FYI according to the [mobile_perf CI job](https://github.com/envoyproxy/envoy/actions/runs/4008291380) this increases the binary size by ~1.4%. Not a blocker for me though, but raising it because reducing binary size have been an active effort in the last few months.

Hm. There's not a lot of new code in this PR, so I can't imagine that's causing a 1% binary size increase. Is this because of the new grpc dependency? I see that the bazelrc is change to only disable grpc in release mode. What config does the mobile_perf job use?

caschoener commented:
> Hm. There's not a lot of new code in this PR, so I can't imagine that's causing a 1% binary size increase. Is this because of the new grpc dependency? I see that the bazelrc is change to only disable grpc in release mode. What config does the mobile_perf job use?

Oh yeah, sorry. I resolved this in chats with some people.  Your guess is correct, mobile_perf was using the test build flags instead of the release ones.  I changed it to use release config in https://github.com/envoyproxy/envoy/pull/25189

caschoener commented:
/retest

caschoener commented:
/retest

caschoener commented:
/retest

caschoener commented:
/retest

caschoener commented:
I changed the sha56 to fix the error 
```
Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/google/libprotobuf-mutator/archive/v1.1.tar.gz] to /build/tmp/_bazel_envoybuild/400406edc57d332f0b9b805d2b8e33a1/external/com_github_google_libprotobuf_mutator/temp10529668633102100589/v1.1.tar.gz: Checksum was 9b862857a725a07287dc2dd182f32b76b38b8129b38aad3e26ebceccd9d6c620 but wanted fd299fd72c5cf664259d9bd43a72cb74dc6a8b9604d107fe2d2e90885aeb7c16
```
Downloaded the file and confirmed it is the new sha56. Maybe we weren't actually downloading this file before, because google_grpc was previously compiled out in tests?

phlax commented on bazel/repository_locations.bzl:
```diff
@@ -309,7 +309,7 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_url = "https://github.com/google/libprotobuf-mutator",
         version = "1.1",
-        sha256 = "fd299fd72c5cf664259d9bd43a72cb74dc6a8b9604d107fe2d2e90885aeb7c16",
+        sha256 = "9b862857a725a07287dc2dd182f32b76b38b8129b38aad3e26ebceccd9d6c620",
```
@caschoener this should not change if the version has not changed

caschoener replied:
I'll change it back.  Was trying to fix a strange checksum error, thought maybe the sha56 was incorrect

phlax replied:
if you require updated deps for your PR - perhaps try and update these independently - that is generally preferable from review/revert pov

caschoener replied:
After talking with JP apparently github tarball checksums aren't stable so this might be a good change if it makes tests pass

phlax replied:
(otherwise - you definitely do not want to change any hashes without changing the corresponding version)

phlax replied:
> After talking with JP apparently github tarball checksums aren't stable so this might be a good change if it makes tests pass

they are stable otherwise we would not be able to hash them

caschoener replied:
You may want to check the envoy mobile collaboration slack channel for discussion on this point

caschoener commented:
/retest

caschoener commented:
/retest

caschoener commented:
/retest

jpsim commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -242,6 +244,23 @@ EngineBuilder& EngineBuilder::enforceTrustChainVerification(bool trust_chain_ver
+#ifndef ENVOY_GOOGLE_GRPC
+  raise std::runtime_error("google_grpc must be enabled in bazel to use ADS")
+#endif
+      custom_ads_ = fmt::format(ads_insert, api_type, address, port);
```
Nit: Indentation?

jpsim commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +539,27 @@ stats_sinks: *stats_sinks
+      rtds_layer:
+        name: {}
+        rtds_config:
+          initial_fetch_timeout:
+            seconds: 1
```
A 1 second timeout seems very aggressive considering real world mobile networking conditions.

I'm not familiar with the implications of this timeout though.

abeyad replied:
agree, this should probably be 5s to start off with.  we can experiment with it as necessary.

caschoener replied:
I'll make it configurable but I think it needs to be 1 second in tests because the tests depend on the initial fetch timing out

jpsim replied:
Why not expose a builder function to configure this then?

caschoener replied:
Just did! Sorry I wasn't clear haha

jpsim replied:
Great, sorry I missed that.

caschoener commented:
Alyssa suggests that I update to make compatible with the new C++ bootstrap builder

abeyad commented:
> Per Alyssa's comments, we'll want to change this to use the proto builder APIs

Ah I see you made the same comment above Chris

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -242,6 +244,23 @@ EngineBuilder& EngineBuilder::enforceTrustChainVerification(bool trust_chain_ver
+                                                            const std::string& address,
+                                                            const int port) {
+#ifndef ENVOY_GOOGLE_GRPC
+  raise std::runtime_error("google_grpc must be enabled in bazel to use ADS");
```
what does `raise` do in C++?  shouldn't this be `throw`?  not familiar with `raise` in c++.

caschoener replied:
Ah good catch

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +539,27 @@ stats_sinks: *stats_sinks
+#{custom_layers}
 )";
+
+const char* rtds_layer_insert = R"(
```
i don't believe we need any of this anymore, right?

abeyad commented on mobile/library/common/config/config.cc:
```diff
@@ -538,5 +539,27 @@ stats_sinks: *stats_sinks
+          resource_api_version: V3
+          ads: {{}})";
+
+const char* ads_insert = R"(
```
same, i don't think we need this anymore, right?

caschoener replied:
Same reply, both builders are currently in use. We are using the yaml approach but verifying it matches against the protobuf.

abeyad commented on mobile/library/cc/engine_builder.cc:
```diff
@@ -366,6 +385,21 @@ std::string EngineBuilder::generateConfigStr() const {
+  if (!rtds_layer_name_.empty() && ads_api_type_.empty()) {
+    throw std::runtime_error("ADS must be configured when using RTDS");
+  }
+  if (!rtds_layer_name_.empty()) {
```
don't think we need L391 to L402 anymore, right?

abeyad commented:
@caschoener You don't want to ever rebase, it loses the commit review history for those who are re-reviewing subsequent changes.  The only time rebase is allowed is if you've forgotten to sign your commit with `-s`, as rewriting that commit is the only way to circumvent that problem.  Otherwise, it should always be a merge instead of a rebase.
